presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java

```diff
@@ -321,14 +323,341 @@ public void testHardCpuLimit()
+        q1.setMemoryUsage(new DataSize(2, BYTE));
+        root.updateGroupsAndProcessQueuedQueries();
+
+        // Completion of q1 should reduce the cached memory usage to 0B.
+        q1.complete();
+
+        // q2 starts running since the current memory usage 0B is less than the limit 1.5B.
+        MockResourceUpdatingQueryExecution q2 = new MockResourceUpdatingQueryExecution();
+        child.run(q2);
+        assertEquals(q2.getState(), RUNNING);
+    }
+
+    /**
+     * A test for correct cpu usage update aggregation and propagation in non-leaf nodes. It uses in a multi
+     * level resource group tree, with non-leaf resource groups having more than one child.
+     */
+    @Test(timeOut = 10_000)
+    public void testRecursiveCpuUsageUpdate()
+    {
+        RootInternalResourceGroup root = new RootInternalResourceGroup("root", (group, export) -> {}, directExecutor());
+        InternalResourceGroup rootChild1 = root.getOrCreateSubGroup("rootChild1");
+        InternalResourceGroup rootChild2 = root.getOrCreateSubGroup("rootChild2");
+        InternalResourceGroup rootChild1Child1 = rootChild1.getOrCreateSubGroup("rootChild1Child1");
+        InternalResourceGroup rootChild1Child2 = rootChild1.getOrCreateSubGroup("rootChild1Child2");
+
+        // Set the same values in all the groups for some configurations
+        List<InternalResourceGroup> groups = ImmutableList.of(root, rootChild1, rootChild2, rootChild1Child1, rootChild1Child2);
+        groups.forEach(group -> {
+            group.setCpuQuotaGenerationMillisPerSecond((new Duration(1, SECONDS)).toMillis());
+            group.setHardConcurrencyLimit(100);
+            group.setMaxQueuedQueries(100);
+        });
+
+        root.setHardCpuLimit(new Duration(8, SECONDS));
+        rootChild1.setHardCpuLimit(new Duration(3, SECONDS));
+        // Setting a higher limit for leaf nodes
+        rootChild1Child1.setHardCpuLimit(new Duration(100, SECONDS));
+        rootChild1Child2.setHardCpuLimit(new Duration(100, SECONDS));
+        rootChild2.setHardCpuLimit(new Duration(100, SECONDS));
+
+        MockResourceUpdatingQueryExecution q1 = new MockResourceUpdatingQueryExecution();
+        MockResourceUpdatingQueryExecution q2 = new MockResourceUpdatingQueryExecution();
+        MockResourceUpdatingQueryExecution q3 = new MockResourceUpdatingQueryExecution();
+
+        rootChild1Child1.run(q1);
+        rootChild1Child2.run(q2);
+        rootChild2.run(q3);
+
+        assertEquals(q1.getState(), RUNNING);
+        assertEquals(q2.getState(), RUNNING);
+        assertEquals(q3.getState(), RUNNING);
+
+        q1.consumeCpuTime(new Duration(2, SECONDS));
+        q2.consumeCpuTime(new Duration(5, SECONDS));
+        q3.consumeCpuTime(new Duration(2, SECONDS));
+
+        // This invocation will update the cached usage for the nodes
+        // root: 9s, rootChild1: 7s, rootChild2:2s, rootChild1Child1: 2s, rootChild1Child2: 5s
```

Comment:
Being able to validate the resource usage snapshot directly here would make this test a lot more readable.